src/specs.ts

```diff
@@ -0,0 +1,78 @@
+export type Feature = {
+    group: string;
+    value: string;
+};
+
+export type Engagement = {
+    content: string | Record<string, unknown>;
+    label: string;
+    boost?: number;
+    features?: Feature[];
+    feed_id?: string;
+    location?: string;
+    position?: number;
+    score?: number;
+};
+
+export type Impression = {
+    content_list: Array<string | Record<string, unknown>>;
+    boost?: number;
+    features?: Feature[];
+    feed_id?: string;
+    location?: string;
+};
+
+const validateFeatures = (features?: Feature[]) => {
+    if (!features) return '';
+    if (!Array.isArray(features)) return 'features should be array';
+
+    for (let i = 0; i < features.length; i += 1) {
```

Comment:
```js
for (let [group, value] of features) {
  if (!group || typeof group !== 'string') ...
  ....
}
```

might be better

Reply:
for..of falls into `no-restricted-syntax` of the linter